publication/models/distribution_list.py

```diff
@@ -145,37 +145,91 @@ def _onchange_keyfields(self):
     @api.model
     def create(self, vals):
         result = super(DistributionList, self).create(vals)
-        result._limit_count()
+        self._update_contract_partner_copies(
+            result.product_id, result.contract_partner_id,
+        )
         return result
 
     @api.multi
     def write(self, vals):
+        old_values = set(self.mapped(
+            lambda x: (x.product_id, x.contract_partner_id)
+        )) if 'product_id' in vals else []
         result = super(DistributionList, self).write(vals)
-        self._limit_count()
+        needs_update = set([
+            'product_id', 'partner_id', 'contract_partner_id', 'copies'
+        ]) & set(vals.keys())
+        if needs_update:
+            for this in self:
+                self._update_contract_partner_copies(
+                    this.product_id, this.contract_partner_id,
+                )
+            for product, partner in old_values:
+                self.env['distribution.list']._update_contract_partner_copies(
+                    product, partner,
+                )
         return result
 
     @api.multi
-    def _limit_count(self):
-        """Limit number of copies send to amount set in contract lines.
+    def unlink(self):
+        updates = set(self.mapped(
+            lambda x: (x.product_id, x.contract_partner_id)
+        ))
+        result = super(DistributionList, self).unlink()
+        for product, partner in updates:
+            self._update_contract_partner_copies(product, partner)
+        return result
 
-        It should be possible to make a constrains method of this function,
-        but for inexplicable reasons this does not work.
+    @api.model
+    def _update_contract_partner_copies(self, product, partner, force=False):
+        """take care that the amount of copies for contract_partner_id partner
+        and product is the same as contracted by this partner for the product.
+        Balance differences by creating/adjusting a distribution.list entry for
+        partner.
+        TODO: if we want to support start and end dates, thing get a bit
```

Comment:
The distribution list does not need any history. So we should basically for the moment only look at contract lines valid on the current date. Then we should have some cronjob to update distribution list entries when start or end dates pass. That would be outside of the scope of this PR I think.